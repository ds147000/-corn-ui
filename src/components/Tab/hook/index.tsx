/* eslint-disable import/export */
/* eslint-disable import/first */

// #if _APP === 'weapp'
export { default as useTabScroll } from './mounted.taro'


// #else
export { default as useTabScroll } from './mounted.h5'
// #endif


