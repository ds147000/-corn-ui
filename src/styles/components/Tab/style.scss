.xrk-tab {
  overflow-y: hidden;
  overflow-x: auto;
  height: 90px;
  white-space: nowrap;

  &::-webkit-scrollbar {
    height: 0;
  }

  &.xrk-tab-button {
    height: 60px;
    justify-content: space-between;
  }

  &-item {
    height: 90px;
    white-space: nowrap;
    line-height: 90px;
    font-size: 32px;
    font-weight: 400;
    color: $text2;
    position: relative;
    flex: 1;
    text-align: center;
    padding: 0 20px;

    &:first-child.active::before {
      transform: translate(calc(-50% - 10px) , 0);
    }

    &:last-child.active::before {
      transform: translate(calc(-50% + 10px) , 0);
    }

    &-active {
      font-size: 38px;
      font-weight: bold;
      color: $text1;
      position: relative;
      z-index: 1;

      &::before {
        content: '';
        background-image: url('../../../assets/img/tab-active.png');
        background-repeat: no-repeat;
        background-size: 100% 100%;
        width: 100px;
        height: 30px;
        position: absolute;
        z-index: -1;
        top: 52px;
        left: 50%;
        transform: translate(-50% , 0);
        contain: layout;
      }
    }

    &-icon {
      height: 90px;

      img {
        width: unset;
      }
    }

    &-message {
      position: relative;
      display: inline-block;

      &::after {
        content: '';
        width: 16px;
        height: 16px;
        background: $urgent;
        border-radius: 16px;
        position: absolute;
        right: -20px;
        top: 10px;
      }
    }
  }
}
