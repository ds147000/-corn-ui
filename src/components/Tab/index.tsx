import React, { useCallback, useRef } from 'react'
import { View } from '@tarojs/components'
import Button from '../Button'
import { useTabScroll } from './hook'
import { Item } from './item'

export namespace TAB {
  export type Change = (index?: number) => void

  export interface Props {
    /** Tab选项数组 */
    options: Item[]
    /** 当前激活索引 */
    currenIndex: number
    /** 风格 */
    type?: 'button' | 'default'
    /** 索引发生改变回调 */
    onChange?(index: number): void
  }

  export interface Item {
    id?: string
    title?: string
    onClick?(): void
    icon?: string
    active?: boolean
    message?: boolean
    url?: string
  }
}


const Tabs: React.FC<TAB.Props> = ({ options, currenIndex, type = 'default', onChange }) => {
  const El = useRef<HTMLDivElement>()
  useTabScroll(El, currenIndex)

  const renderItem = useCallback((item: TAB.Item, active: boolean, _onChange: TAB.Change): JSX.Element => {
    const key = item.id || item.title || item.icon

    if (type === 'button') {
      return (
        <Button
          type={active ? 'primary' : 'light'}
          onClick={_onChange}
          key={key}
          className={active ? '__active' : ''}
        >
          {item.title}
        </Button>
      )
    }

    return (
      <Item
        {...item}
        active={active}
        key={key}
        onClick={_onChange}
      />
    )
  }, [ type ])

  return (
    <View ref={El} className={`xrk-tab xrk-f xrk-tab-${type}`} data-testid="tab" >
      {options?.map((item, key) => renderItem(item, currenIndex === key, (): void => onChange?.(key)))}
    </View>
  )
}

export default Tabs
