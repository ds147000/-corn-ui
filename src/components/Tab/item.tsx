import React, { useMemo } from 'react'
import ClassNames from 'classnames'
import { Text, View } from '@tarojs/components'
import Image from '../Image'
import type { TabItemPorps } from './index'

export const Item: React.FC<TabItemPorps> = ({ icon, active, title, onClick, message }) => {

  const _class = useMemo(() => ClassNames(
    'xrk-tab-item __x-st-exp __x-st-click',
    active && 'xrk-tab-item-active'
  ), [ active ])

  return (
    <View className={_class} onClick={onClick} data-testid="tab-item" >
      <Text className={message ? 'xrk-tab-item-message' : ''}>
        {icon ? <Image src={icon} className="xrk-tab-item-icon" imgProps={{ alt: title }} /> : title}
      </Text>
    </View>
  )
}
