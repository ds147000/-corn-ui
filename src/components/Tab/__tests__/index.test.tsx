/* eslint-disable no-magic-numbers */
import React, { useState } from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import Tab, { TAB } from '../index'

jest.mock('../hook', () => ({
  useTabScroll: jest.requireActual('../hook/mounted.h5.tsx').default
}))

jest.mock('../../../utils', () => ({
  getRanDomId: () => '123'
}))


test('option', () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const screen = render(<Tab options={options} currenIndex={1} />)
  options.map((item) => screen.getByText(item.title as string))
  expect(screen.container).toMatchSnapshot()
})

test('option middle', () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const screen = render(<Tab options={options} currenIndex={1} size="middle" />)
  options.map((item) => screen.getByText(item.title as string))
  expect(screen.container).toMatchSnapshot()
})


test('option of id', () => {
  const options: TAB.Item[] = [
    { title: 'item1', id: '03' },
    { title: 'item1', id: '02' },
    { title: 'item1', id: '01' },
    { title: 'item22121' }
  ]

  const screen = render(<Tab options={options} currenIndex={1} />)
  expect(screen.container).toMatchSnapshot()
})


test('option of button', () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const screen = render(<Tab options={options} currenIndex={1} type="button" />)
  options.map((item) => screen.getByText(item.title as string))
  expect(screen.container).toMatchSnapshot()
})

test('active', () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const screen = render(<Tab options={options} currenIndex={1} />)
  options.map((item) => screen.getByText(item.title as string))
  expect(screen.getByText('item2').parentElement?.className.indexOf('active')).not.toBe(-1)
})

test('onChange', async () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const App: React.FC<unknown> = () => {
    const [ current, setCurrent ] = useState(0)

    const onChange = (index: number): void => setCurrent(index)
    return (
      <Tab options={options} currenIndex={current} onChange={onChange} />
    )
  }

  const screen = render(<App />)
  expect(screen.getByText('item1').parentElement?.className.indexOf('active')).not.toBe(-1)
  fireEvent.click(screen.getByText('item2'))
  await waitFor(() => expect(screen.getByText('item2').parentElement?.className.indexOf('active')).not.toBe(-1))
})

test('onChange of ios', async () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const App: React.FC<unknown> = () => {
    const [ current, setCurrent ] = useState(0)

    const onChange = (index: number): void => setCurrent(index)
    return (
      <Tab options={options} currenIndex={current} onChange={onChange} />
    )
  }

  const screen = render(<App />)
  // eslint-disable-next-line @typescript-eslint/no-shadow, @typescript-eslint/no-explicit-any
  screen.getByTestId('tab').scrollTo = (options: unknown): any => {
    if (typeof options === 'object') throw 'error'
    return options
  }
  expect(screen.getByText('item1').parentElement?.className.indexOf('active')).not.toBe(-1)
  fireEvent.click(screen.getByText('item2'))
  await waitFor(() => expect(screen.getByText('item2').parentElement?.className.indexOf('active')).not.toBe(-1))
})

test('onChange of button', async () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const App: React.FC<unknown> = () => {
    const [ current, setCurrent ] = useState(0)

    const onChange = (index: number): void => setCurrent(index)
    return (
      <Tab options={options} currenIndex={current} onChange={onChange} type="button" />
    )
  }

  const screen = render(<App />)

  fireEvent.click(screen.getByText('item2'))
})

test('onChange of error index', async () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { title: 'item4' }
  ]

  const App: React.FC<unknown> = () => {
    const [ current, setCurrent ] = useState(0)

    const onChange = (): void => setCurrent(100)
    return (
      <Tab options={options} currenIndex={current} onChange={onChange} />
    )
  }

  const screen = render(<App />)
  fireEvent.click(screen.getByText('item2'))
  await waitFor(() => expect(screen.getByText('item2').parentElement?.className.indexOf('active')).toBe(-1))
})

test('icon', () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { icon: '2121.png' }
  ]

  const screen = render(<Tab options={options} currenIndex={1} />)
  expect(screen.container).toMatchSnapshot()
})

test('not currenIndex', () => {
  const options: TAB.Item[] = [
    { title: 'item1' },
    { title: 'item2' },
    { icon: '2121.png' }
  ]

  const screen = render(<Tab options={options} currenIndex={-1} />)
  expect(screen.container).toMatchSnapshot()
})
