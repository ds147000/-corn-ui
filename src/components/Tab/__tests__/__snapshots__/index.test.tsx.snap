// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon 1`] = `
<div>
  <div
    class="xrk-tab xrk-f xrk-tab-default"
    data-testid="tab"
    id="123"
  >
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item1
      </div>
    </div>
    <div
      class="xrk-tab-item __x-st-exp __x-st-click xrk-tab-item-active __active"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item2
      </div>
    </div>
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        <div
          class="taro-img xrk-img xrk-tab-item-icon"
        >
          <img
            class="taro-img__mode-scaletofill"
            src="2121.png"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`not currenIndex 1`] = `
<div>
  <div
    class="xrk-tab xrk-f xrk-tab-default"
    data-testid="tab"
    id="123"
  >
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item1
      </div>
    </div>
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item2
      </div>
    </div>
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        <div
          class="taro-img xrk-img xrk-tab-item-icon"
        >
          <img
            class="taro-img__mode-scaletofill"
            src="2121.png"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`option 1`] = `
<div>
  <div
    class="xrk-tab xrk-f xrk-tab-default"
    data-testid="tab"
    id="123"
  >
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item1
      </div>
    </div>
    <div
      class="xrk-tab-item __x-st-exp __x-st-click xrk-tab-item-active __active"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item2
      </div>
    </div>
    <div
      class="xrk-tab-item __x-st-exp __x-st-click"
      data-testid="tab-item"
    >
      <div
        class=""
      >
        item4
      </div>
    </div>
  </div>
</div>
`;

exports[`option of button 1`] = `
<div>
  <div
    class="xrk-tab xrk-f xrk-tab-button"
    data-testid="tab"
    id="123"
  >
    <div
      class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-light"
    >
      item1
    </div>
    <div
      class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-primary __active"
    >
      item2
    </div>
    <div
      class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-light"
    >
      item4
    </div>
  </div>
</div>
`;
